src/masternodes/rpc_evm.cpp

```diff
@@ -220,11 +222,57 @@ UniValue vmmap(const JSONRPCRequest &request) {
 
     const std::string input = request.params[0].get_str();
 
-    const int typeInt = request.params[1].get_int();
+    int typeInt = request.params[1].get_int();
     if (typeInt < 0 || typeInt >= VMDomainRPCMapTypeCount) {
         throwInvalidParam();
     }
-    const auto type = static_cast<VMDomainRPCMapType>(request.params[1].get_int());
+
+    // auto infers type
+    if (typeInt == 0) {
+        // block number
+        uint64_t height;
+        if (ParseUInt64(input, &height)) {
+            CrossBoundaryResult result;
+            auto evmBlockCount = evm_try_get_block_count(result);
+            if (!result.ok) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, result.reason.c_str());
+            }
+            // evm block count always less than dvm block count
+            if (height > evmBlockCount) {
+                typeInt = 5; // BlockNumberDVMToEVM
+            } else {
+                // dvm input(after evm enabled) always greater than evm block count
+                typeInt = 6; // BlockNumberEVMToDVM
+            }
+        } else if (input.length() == 64 || input.length() == 66) { // hash
+            ResVal res = ResVal<uint256>(uint256{}, Res::Ok());
+            res = pcustomcsview->GetVMDomainTxEdge(VMDomainEdge::DVMToEVM, uint256S(input));
+            if (!res) {
+                res = pcustomcsview->GetVMDomainTxEdge(VMDomainEdge::EVMToDVM, uint256S(input));
+                if (!res) {
+                    res = pcustomcsview->GetVMDomainBlockEdge(VMDomainEdge::DVMToEVM, uint256S(input));
+                    if (!res) {
+                        res = pcustomcsview->GetVMDomainBlockEdge(VMDomainEdge::EVMToDVM, uint256S(input));
+                        if (!res) {
+                            throw JSONRPCError(RPC_INVALID_PARAMETER, "Unsupported type or unable to determine conversion type automatically from the input");
+                        } else {
+                            typeInt = 4;
+                        }
+                    } else {
+                        typeInt = 3;
+                    }
+                } else {
+                    typeInt = 2;
+                }
+            } else {
+                typeInt = 1;
+            }
```

Comment:
We can move this into a function that returns early when it matches successfully, and throw once at the end if none of them match.

Reply:
Yeah looks good, tho I wonder if we even need the `res` variable. It could work by just checking `GetVMDomainTxEdge` in the `if` statement.

```cpp
    if (pcustomcsview->GetVMDomainTxEdge(VMDomainEdge::DVMToEVM, uint256S(input))) {
        return 1; // VMDomainRPCMapType::TxHashDVMToEVM
    }
```

Reply:
We can also construct `uint256S(input)` just once as `const` to prevent duplicate calculations.

Reply:
resolved both at each
https://github.com/DeFiCh/ain/pull/2267/commits/8e13937731ea09468ddc7e5e4211420670cfcf7f
https://github.com/DeFiCh/ain/pull/2267/commits/58dd53c827fa9c318950c910b8186a8abfe83965